02cpp1/sec03Pointers.md

```diff
@@ -0,0 +1,364 @@
+
+## Memory Ownership
+
+Smart pointers are designed to model the ownership of memory allocated on the heap. **You should never create smart pointers which point to variables already declared on the stack.** Objects declared on the stack are destroyed when the original variable goes out of scope, and it is not possible to transfer this ownership. Therefore, any smart pointer pointing to a stack object *cannot* properly model the ownership of the data. If the original variable goes out of scope first, then your pointers will be left dangling. If you pointer goes out of scope first, then your program will crash even if the original variable isn't accessed again, because the smart pointer will not have the right to delete that data. If you need to point to stack data, you will need to use a *raw pointer*, which does not embody any ownership.  
+
+## Raw Pointers
+
+You may have come across *raw pointers* before when using C++. These have existed in C++ for longer (indeed, they pre-date C++), and work slightly differently to smart pointers. The most important difference is that **raw pointers don't do any automatic memory management**. This means that if you use a raw pointer to assign some memory, and you do not manually free that memory before the pointer goes out of scope, that memory will become inaccessible and will not be freed until your program terminates. This is called a memory leak, and can be extremely important when running large programs, where you risk running out of memory and therefore prematurely aborting your program execution. 
+
+**Avoid using raw pointers for objects whose memory is not owned by some other resource** (e.g. a stack variable, a container, or a smart pointer). Raw pointers can be used for "non-owning pointers", which have no impact on the lifetime of the object that they point to, **as long as you can be sure that they won't point to invalid memory**. This can be useful for graph like classes where objects can reference one another (including circular references) but do not influence the lifetime of each other, or for referencing objects in container classes like `vector`. Bear in mind however, that we can't check if a raw pointer points to valid memory of not, so if the object that the raw pointer points to is deleted before the raw pointer, then accessing that pointer will cause undefined behaviour.
+- Okay for non-owning pointers 
+- Raw pointers should be used when you need to point to an existing stack variable. **Never point to an existing stack variable using smart pointers as their memory management will conflict.**
+- Faster memory access than weak pointers, although unique/shared pointers can match speed of raw pointers if compiled with optimisation.  
```

Comment:
```suggestion
- Faster memory access than weak pointers, although unique/shared pointers will match speed of raw pointers if compiled with optimisation.  
```

I don't want to give students any ammunition to claim performance reasons for not using smart pointers.